fiona/model.py

```diff
@@ -17,31 +53,269 @@ class Object(MutableMapping):
     about future deprecation of features.
     """
 
-    def __init__(self, **data):
-        self._data = {}
-        self._data.update(data)
+    _delegated_properties = []
+
+    def __init__(self, **kwds):
+        self._data = OrderedDict(**kwds)
+
+    def _props(self):
+        return {
+            k: getattr(self._delegate, k)
+            for k in self._delegated_properties
+            if k is not None
+        }
 
     def __getitem__(self, item):
-        return self._data[item]
+        props = self._props()
+        props.update(**self._data)
+        return props[item]
 
     def __iter__(self):
-        return iter(self._data)
+        props = self._props()
+        return itertools.chain(iter(props), iter(self._data))
 
     def __len__(self):
-        return len(self._data)
+        props = self._props()
+        return len(props) + len(self._data)
 
     def __setitem__(self, key, value):
         warn(
             "instances of this class -- CRS, geometry, and feature objects -- will become immutable in fiona version 2.0",
             FionaDeprecationWarning,
             stacklevel=2,
         )
-        self._data[key] = value
+        if key in self._delegated_properties:
+            setattr(self._delegate, key, value)
+        else:
+            self._data[key] = value
 
     def __delitem__(self, key):
         warn(
             "instances of this class -- CRS, geometry, and feature objects -- will become immutable in fiona version 2.0",
             FionaDeprecationWarning,
             stacklevel=2,
         )
-        del self._data[key]
+        if key in self._delegated_properties:
+            setattr(self._delegate, key, None)
+        else:
+            del self._data[key]
+
+
+class _Geometry(object):
+    def __init__(self, coordinates=None, type=None):
+        self.coordinates = coordinates
+        self.type = type
+
+
+class Geometry(Object):
+    """A GeoJSON-like geometry
+
+    Notes
+    -----
+    Delegates coordinates and type properties to an instance of
+    _Geometry, which will become an extension class in Fiona 2.0.
```

Comment:
Could you explain the purpose of these delegation classes? Why not expose the extension class as `Geometry` in the future?

Reply:
@snorfalorpagus it's a compromise. On the one hand we need an extension class to get the performance boost we're looking for. On the other hand, using only an extension class limits our implementation options. A Cython cdef class can't derive from a Python class, so we can't inherit MutableMapping and gain its methods.

The delegation is working out well in Rasterio's CRS case, and I think we could benefit from it here.